### PR TITLE
test: testing ability to upgrade after trapping finally

### PR DESCRIPTION
**Pull request by crusso**
A trapping finally block will (necessarily) leak a callback table slot.

Motoko prevents a canister with non-empty callback table from upgrading, but only unless the canister is stopped.

This test verifies that we can recover from a callback table leak (due a faulty finally) by first stopping the canister, upgrading and restarting it.

This PR also adds an admonition documenting the danger of trapping finally-blocks and how to recover from them.

### CODE REVIEW

File: test/run-drun/try-finally-trap.drun [hunk 1]
```diff
@@ -0,0 +1,6 @@
+install $ID try-finally-trap/try-finally-trap.mo ""
+upgrade $ID try-finally-trap/try-finally-trap.mo ""
+ingress $ID go "DIDL\x00\x00"
+ingress $ID show "DIDL\x00\x00"
+upgrade $ID try-finally-trap/try-finally-trap.mo ""
+ingress $ID show "DIDL\x00\x00"
```
Comment (crusso) [inline test/run-drun/try-finally-trap.drun:6]:
```suggestion
ingress $ID show "DIDL\x00\x00"

```

File: test/run-drun/try-finally-trap-stop.mo [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import Prim "mo:⛔";
+
+// tests that a canister with a leaky callback table
+// (due to a trapping finally block) preventing upgrades,
+// can still be upgraded after being stopped.
+actor {
+
+  let ic00 = actor "aaaaa-aa" : actor {
+    stop_canister : { canister_id : Principal} -> async ();
+    start_canister : { canister_id : Principal} -> async ();
```
Comment (ggreif) [inline test/run-drun/try-finally-trap-stop.mo:11]:
```suggestion
    stop_canister : { canister_id : Principal } -> async ();
    start_canister : { canister_id : Principal } -> async ();
```

File: test/run-drun/try-finally-trap/try-finally-trap.mo [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import { debugPrint; error; call_raw; principalOfActor; rts_callback_table_count } =  "mo:⛔";
+
+actor A {
+
+  public func go() : async () {
+    try {
+      await async ();
+      assert false
```
Comment (ggreif) [inline test/run-drun/try-finally-trap/try-finally-trap.mo:8]: this isn't needed. There is a trap coming up.

File: test/run-drun/try-finally-trap/try-finally-trap.mo [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    }
+    finally {
+      debugPrint("trap in finally!");
```
Comment (ggreif) [inline test/run-drun/try-finally-trap/try-finally-trap.mo:11]: Then I expect to see this output twice.

**Review by ggreif — APPROVED** [review]
LGTM! Please have a look at the comments.

File: test/run-drun/try-finally-trap/try-finally-trap.mo (hunk 4 above)
Comment (crusso) [reply test/run-drun/try-finally-trap/try-finally-trap.mo:11]: Don't follow, why?

File: test/run-drun/try-finally-trap/try-finally-trap.mo (hunk 3 above)
Comment (crusso) [reply test/run-drun/try-finally-trap/try-finally-trap.mo:8]: Right you are.

File: doc/md/reference/language-manual.md [hunk 5]
```diff
@@ -2550,11 +2550,22 @@ Because the [`Error`](../base/Error.md) type is opaque, the pattern match cannot
...
 The preceding `catch` clause may be omitted in the presence of a `finally` clause.
 
 This form is `try <block-or-exp1> (catch <pat> <block-or-exp2>)? finally <block-or-exp3>`, and evaluation proceeds as above with the crucial addition that every control-flow path leaving `<block-or-exp1>` or `<block-or-exp2>` will execute the unit-valued `<block-or-exp3>` before the entire `try` expression produces its result. The cleanup expression will additionally also be executed when the processing after an intervening `await` (directly, or indirectly as `await*`) traps.
 
+:::danger
+
+The code within a `finally` block should terminate promptly and not trap.
+A trapping finally block will fail to free its callback table slot which
+can prevent a future upgrade.
+In this situation, the canister should be explicitly stopped before re-attempting the upgrade.
```
Comment (crusso) [inline doc/md/reference/language-manual.md:2563]:
```suggestion
In this situation, the canister should be explicitly stopped before re-attempting the upgrade.
In addition, care should be taken to release any resources that may have remained acquired due to the trap.
```

File: test/run-drun/try-finally-trap/try-finally-trap.mo (hunk 4 above)
Comment (crusso) [reply test/run-drun/try-finally-trap/try-finally-trap.mo:11]: OMG, you are right. But that's weird.  I guess we execute the finally on the success path, it traps and rolls back and then executes the finally a second time in the cleanup. I guess that's benign except for the logging....

Comment (ggreif) [reply test/run-drun/try-finally-trap/try-finally-trap.mo:11]: Right. Nothing to be ashamed of...

**Pull request merged**